### ISSUE TITLE
Implementation of SemCKD

### ISSUE THREAD

**Issue by AndyFrancesco29** [yoshitomo-matsubara/torchdistill#140]
Hi @yoshitomo-matsubara ,

Thanks for your advice before. Right now I have implemented another method [SemCKD](https://github.com/DefangChen/SemCKD) based on this framework. Due to the way it trains the student I have made some changes on the framework, but I am no sure if this could hurt the pipeline.

1. In this method when the student network forward, it needs to reshape the output feature of each layer to the same one as feature in teacher network, so we need to know the shape of feature in each layer in teacher network. In order to do that I add a new forward_proc function like this:

```
@register_forward_proc_func
def forward_batch_teacher_output(model, sample_batch, targets=None, supp_dict=None, teacher_io_dict=None):
    return model(sample_batch, teacher_io_dict)
```

I am not sure if there is a better way to do that.

2. Also I also designed a wrapper function like SSKD to calculate module aside from the original backbone such as key, query and value in attention etc. The problem is that in the old frame it extract every value with hook manager including those in post_forward function, so it could cause a long list of path in config file. I add a return value in post forward, so I put all those value in post_forward into a tuple and input to loss module directly. Maybe using hook manager only in original backbone might be easier to transfer different codes?

3. Also there is another question here: I found there are two losses: org_term and sub_term. So if I only use one kind of loss (such as SemCKD) they are the same right? If I want to implement SemCKD+CRD, then I can directly use these two terms?

Thank you very much!

**Comment by yoshitomo-matsubara** [comment]
Hi @AndyFrancesco29 

I have not carefully read the paper of SemCKD yet, but the following answers should resolve your concerns.

## 1 and 2.
With torchdistill, the easiest way to implement such methods is to use `SpecialModule`, a kind of wrapper of either teacher or student models, can be used as a local teacher/student model. e.g., If you're familiar with the [Correlation Congruence for Knowledge Distillation](https://openaccess.thecvf.com/content_ICCV_2019/papers/Peng_Correlation_Congruence_for_Knowledge_Distillation_ICCV_2019_paper.pdf), [this sample config file](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/ilsvrc2012/single_stage/cckd/resnet18_from_resnet50.yaml#L70-L100) and [this SpecialModule](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/torchdistill/models/special.py#L252-L280) would help you.
1. SpecialModule wraps your teacher/student model, and [its forward function](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/torchdistill/models/special.py#L272-L276) is used like your original teacher/student model. Once the forward function is called, the teacher_io_dict and/or student_io_dict will be updated.
2. Then, [its post_forward function](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/torchdistill/models/special.py#L278-L280) will be called to [get an input or output from a hooked module specified in the yaml file](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/torchdistill/models/special.py#L279) and [feed it to a module newly introduced to the wrapper](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/torchdistill/models/special.py#L280).
3. [Don't forget to register a hook to the new module](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/ilsvrc2012/single_stage/cckd/resnet18_from_resnet50.yaml#L84) (here, `linear`) to be used in [the subsequent loss module](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/torchdistill/losses/single.py#L426-L427)).

> The problem is that in the old frame it extract every value with hook manager including those in post_forward function, so it could cause a long list of path in config file. 

If you avoid writing a long list of path in config file, you'll need to *hardcode* such paths in modules and change a lot of interfaces in torchdistill pipeline. It means that you'll need to prepare such a module class for each of the models you train (e.g., one module class with hardcoded paths for ResNet, the other with hardcoded paths for DenseNet, and so on). Like you see in many of knowledge distillation repositories, such modules are not implemented in a model-agnostic way, thus difficult to be applied to different models. Addressing such issues is one of the core motivations I developed and published torchdistill.


> Also there is another question here: I found there are two losses: org_term and sub_term. So if I only use one kind of loss (such as SemCKD) they are the same right? If I want to implement SemCKD+CRD, then I can directly use these two terms?

`org_term` is used for loss module that can be defined simply with student's final output and target (+ potentially, teacher's final output) such as CrossEntropyLoss (student final output and target) and KLDiv between final outputs from student and teacher models. `sub_terms` are everything else e.g., MSE losses between intermediate outputs from the models.
If it looks like SemCKD uses a linear combination of `CrossEntropyLoss` and `SemCKDLoss`, you'll define the former and latter under `org_term` and `sub_terms` respectively.
By the way, personally I'd not suggest you use CRD even as additional loss as it's very time- and resource-consuming approach as demonstrated in [my paper](https://github.com/yoshitomo-matsubara/torchdistill#citation).

**Comment by yoshitomo-matsubara** [comment]
Hi @AndyFrancesco29 
Thank you for the clarification and suggestion.

> What I mean is that suppose there is a new method, and with your framework, I could only focus on build a wrapper (SpecialModule), a loss and a yaml file. In the wrapper, the model can be seen as two parts: a backbone and a method-specific part. For example in SSKD method, the backbone is Resnet and Densenet while SS module is the method-specific part.

Yes, your understanding sounds correct.

> I agree your way to extract feature from backbone with hook manager, but I am not sure if using hook manager to get output from method-specific part is necessary. If you take a look at the SemCKD, you can find that this part generates many outputs that is useful for loss calculation. So I simply return these part from post_forward method. In SSKD method, it will simply return the output of SS module. Since it is method-specific part, it is less likely that it will be reused in another method. Since we have to write specific wrapper for this method so I do not think it is a hard code problem. Since the hook manager only extract feature from commonly used models, it will reduce the parameters used in yaml.

What I meant is that I'd avoid hardcoding ***model-specific*** things (being method-specific should be fine as each SpecialModule is designed for a specific method).
SemCKD's implementation is more or less the same as RepDistiller since their code is based on the repo, which is model-specific and I'd avoid.
Your suggested approach may reduce parameters in yaml file, which is good, but at the same time it will make the method black-box like which layers we need to extract input/output, which loss function uses which pair of modules's input/output, etc.


> Also in some occasion we need to use teacher output when forward or post forward student network, and in the framework I think two networks do not communicate with each other until loss calculation, that's why I design a forward_proc method at the first.

In my opinion, ultimately, there should be no communication between teacher and student models until loss function since I could not come up with a valid reason that both the models need to communicate each other except at loss calculation.

Overall, I appreciate your feedback. I am not very obsessed with the current SpecialModule approach, and there may be a better approach for torchdistill. But at this point I think that the current approach is more suitable for torchdistill.
(Needless to say, you're free to take that approach for your use.)

**Comment by AndyFrancesco29** [comment]
Hi @yoshitomo-matsubara ,

Thanks for your reply.

> As described in my paper, making yaml config files look high-level summary of experiments is also a goal of torchdistill, and it will help people understand the difference between methods at high-level, by comparing the yaml files.

I have read your paper again and I understood why you design this framework in this way and how your framework is different from Repdistiller. I have to say I did not get your point at the first time. I am still working on SemCKD to find a better implementation under this framework. If you have time, I would be really glad if you could implement this method since it is a novel and effective method proposed this year.

Overall I appreciate your effort on this framework. It helps me a lot when I do research on knowledge distillation area. And I am grateful for your patience (I really have a lot of questions lol). Thank you very much for that.

**Comment by yoshitomo-matsubara** [comment]
Hi @AndyFrancesco29 

Great to hear that torchdistill helped your research on KD :)
I'm planning to add [knowledge review method](https://openaccess.thecvf.com/content/CVPR2021/html/Chen_Distilling_Knowledge_via_Knowledge_Review_CVPR_2021_paper.html) soon, and then will take a look at SemCKD to see if it's easy to implement. I'll mention this issue when creating the PR.

For additional questions, feel free to reach out to me at [the discussion forum](https://github.com/yoshitomo-matsubara/torchdistill/discussions). As mentioned in README, I prefer keeping questions, feature request, and so on in the discussion page unless they come with some errors from torchdistill.

**Comment by AndyFrancesco29** [comment]
Hi @yoshitomo-matsubara ,

Sure I will close this issue and use discussion in the future. I am also implementing this review KD method right now. Maybe I could compare my implementation with yours later. Thanks a lot!

**Comment by yoshitomo-matsubara** [comment]
FYI, https://github.com/yoshitomo-matsubara/torchdistill/pull/141 is the PR for the knowledge review method
